modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -1291,6 +1302,17 @@ void ONNXImporter::parseReduce(LayerParams& layerParams, const opencv_onnx::Node
     layerParams.type = (depth == CV_8S) ? "ReshapeInt8" : "Reshape";
     layerParams.set("dim", DictValue::arrayInt(&targetShape[0], targetShape.size()));
 
+    // Set batchsize dim as dynamic to be compatible with batch size >= 2.
+    if (targetShape[0] == 1 && targetShape.size() > 1)
+    {
+        std::vector<int> dynamicAxes = {0};  // The index of batchsize dim is 0.
+        std::vector<int> inputIndices = {0};
+
+        layerParams.set("has_dynamic_shapes", true);
+        layerParams.set("dynamic_axes", DictValue::arrayInt(dynamicAxes.data(), dynamicAxes.size()));
+        layerParams.set("input_indices", DictValue::arrayInt(inputIndices.data(), inputIndices.size()));
+    }
```

Comment:
Thanks for code reviewing @rogday. This part code is for dynamic batch size. In the current implementation, the `parseReduce` consists of `ReduceLayer` and `ReshapeLayer`.
And the `ReduceLayer` supports dynamic batch size by default. And for `ReshapeLayer` to support dynamic batch size needs to manually specify the index of the dynamic dimension.

So, I check `if targetShape.size() > 1` first, and then set the index `0` as dynamic shape.
Maybe we can change the judgment logic of `if` at line 1306 or directly do not support dynamic batch size in `parseReduce`.

Reply:
For example, the input shape is [1, 4, 4, 4]. And the ONNX model one layer of ReduceSum with axis 2, keepdim = 1. But the shape of input blob is [2, 4, 4, 4]. In parseReduce, the ReduceSum layer can output shape with [2, 4, 4]. But the Reshape output is set to [1, 4, 1, 4]. In this case, if we set the index 0 as dynamic shape, the Reshape Layer will re-compute the output shape and output the [2, 4, 1, 4].

BTW, we use Reshape Layer in the parseReduce because when keepdim = 1, we need to reserve the dimension of shape 1.

Reply:
Hi @rogday, you're right. If the user has a dynamic shape at batchsize dimension, then the input shape should be [0, 4, 4, 4]. To my knowledge, if the input shape of model is [1, 4, 4, 4] but the shape of the given input is [2, 4, 4, 4], we should support such batchsize dynamic shape (or multi-batch size). 
In order to check this, I have generated two different dynamic shape regression test:
1. [1, 4, 4, 4] [reduce_sum_axis_dynamic_batch](https://github.com/opencv/opencv_extra/pull/987).
2.  [?, 4, 4, 4] [google drive](https://drive.google.com/drive/folders/1CtoF7iWUf5VvY82fK9IjqMH1sHuAN3Rm?usp=sharing)

Reply:
Hi @rogday, any update?

Reply:
I wanted to check if onnxruntime supports this, but couldn't validate since the models in the test are invalid:

`
InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : Load model from reduce_sum_axis.onnx failed:This is an invalid model. Type Error: Type 'tensor(float)' of input parameter (axis_tensor) of operator (ReduceSum) in node () is invalid.
`

`axis_tensor` parameter should be of different type according to the ONNX [spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#reducesum).


And even if we fix the test(use int64 instead of float32 for axes tensor), onnxruntime will disagree with us if we try to execute with different-shaped input:

`InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input: input1 for the following indices
 index: 0 Got: 2 Expected: 1
 Please fix either the inputs or the model.`
 
 I strongly believe it should be an error in OpenCV as well.

Reply:
Hi @rogday, Thanks for the careful inspection.
OK, I agree. And we only support the second dynamic batch size(["?", 4, 4, 4]) in ReduceSum. And we error the first dynamic batch size (expected input shape is [1, 4, 4, 4], but given [2, 4, 4, 4]).

The code and test case have been updated.

Reply:
Hi @zihaomu, I fed tensor of shape 2x3x4x2 to the `reduce_sum.onnx` which expects 1x3x4x2 and it didn't error for me with your PR. I think we need to dig deeper to address this issue - take a look [here](https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/onnx/onnx_importer.cpp#L869-L872).
If I comment out this if, I'll get 1x3x4x2 in case of `reduce_sum.onnx` and 0x4x4x4 in case of `reduce_sum_axis_dynamic_batch.onnx`. And it will allow you to fill dynamic axes correctly(just collect the ones that are 0). But I don't know the implications of this commenting - we should run all the test that we have(layer and full network tests too).
Current implementation seem like a workaround for me.

Reply:
Do you mean the comment of `FIXIT result is not reliable for models with multiple inputs.`?
I think [this code](https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/onnx/onnx_importer.cpp#L869-L872) is OK for now. It can make the DNN model have a certain batch dynamic shape ability.  But when the model has two inputs and different batch sizes are given by the user, this will lead to unknown errors.

Reply:
Yes, I meant commenting out `inpShape[0] = std::max(inpShape[0], 1);`, testing if something breaks(non of the layer tests failed for me, but I didn't run full networks) and deciding if the problem lies elsewhere. Do you have any other proposals in mind? Because the current solution treats any tensor with shape (1, ...) as dynamic, which is wrong.

Reply:
Hi @rogday, I got you. I commented on the code you recommended and the `Test_ONNX_layers.LSTM_Activations` failed on my computer. The `LSTM_Activations` has a dynamic input and contained a `Reshape layer`. If we want this network work without the code `inpShape[0] = std::max(inpShape[0], 1);`, we need move `inpShape[0] = std::max(inpShape[0], 1);` to `ReshapeLayerImpl::getMemoryShapes`. And maybe there are more layers need this code.

In our current OpenCV DNN structure, we cannot move out of `inpShape[0] = std::max(inpShape[0], 1);` code. 
The reasons are as follows: in the model load phase (`net = readNet(path);`), we will calculate the memory required for each layer. And the dynamic input like [0, 4, 4, 4]  will cause memory calculation errors. In the `net.setInput(blob);` stage, we will update the memory size required by each layer again and then perform the actual memory allocation.

So, we actually have two computations that require memory size and one memory allocation. If we can do only parsing layers in `readNet` stage without memory size calculations, we can remove this code. But this is impossible, because the parsing of some layers needs to know the size of his input.

Reply:
> Do you have any other proposals in mind? 

Actually, we can do the Reshape Layer's things in Reduce Layer's `forward`. That will take some time.

Reply:
Hi @rogday, I have updated the code. And we can support `Reduce Layer` without `Reshape Layer`. I think this is the good option.

Reply:
Hi @rogday, since currently [0, 4, 4, 4] will be parsed as [1, 4, 4, 4], that's why I write that code. Do you have any better suggestions for this? Maybe we should not support dynamic batch size in this PR. After you successfully remove [this part](https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/onnx/onnx_importer.cpp#L869-L872), we will come back to support it.

> I would suggest removing hasDynamicShapes altogether.

I agree. This is a big change, we can do it in another PR.

Reply:
Hi @rogday, I have another interesting discovery. In ONNX, if we have a model with fixed shape like [1, 4, 4, 4], and we give the model a input which shape is [2, 4, 4, 4], it should error in the shape matching stage. But for now, OpenCV only gives this type errors inside some layers' `getMemoryShapes`. I mean it should error earlier if the input shape is not correctly.

Reply:
@zihaomu, I propose to discuss it further on weekly meeting. I agree, we should fail earlier, but sanity checks also come in handy sometimes.

Reply:
Agree.